File: test/src/main/java/com/shipping/service/SystemService.java
package com.shipping.service;

import com.shipping.domain.dto.MenuDto;
import com.shipping.domain.dto.RegionDto;
import com.shipping.domain.enums.FunctionTypeE;
import com.shipping.domain.sys.Function;
import com.shipping.repository.sys.FunctionRepository;
import com.shipping.repository.sys.RegionRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.spin.core.collection.FixedVector;
import org.spin.core.session.SessionManager;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * <p>Created by xuweinan on 2017/8/8.</p>
 *
 * @author xuweinan
 */
@Service
public class SystemService {
    private static final Logger logger = LoggerFactory.getLogger(SystemService.class);

    @Autowired
    private FunctionRepository functionDao;

    @Autowired
    private RegionRepository regionDao;

    private UserService userService;

    /**
     * 组装菜单树
     */
    public List<MenuDto> getMenus() {
        final List<MenuDto> result = new ArrayList<>();

        Map<Integer, List<MenuDto>> group = userService.getUserFunctions(SessionManager.getCurrentUser().getId()).get(FunctionTypeE.MEMU).stream()
            .map(MenuDto::toDto).collect(Collectors.groupingBy(f -> f.getIdPath().split(",").length));

        int maxDept = group.keySet().stream().max(Integer::compareTo).orElse(0), currDept = 1;
        FixedVector<List<MenuDto>> parent = new FixedVector<>(1);

        while (currDept <= maxDept) {
            group.get(currDept++).stream().collect(Collectors.toMap(MenuDto::getIdPath, f -> f)).forEach((k, v) -> {
                parent.put(result);
                Arrays.stream(k.split(",")).map(Long::valueOf).forEach(id -> {
                    Optional<MenuDto> p = parent.get().stream().filter(m -> id.equals(m.getId())).findFirst();
                    if (p.isPresent()) {
                        parent.put(p.get().getChildren());
                    } else {
                        parent.get().add(v);
                    }
                });
            });
        }
        return result;
    }

    public List<RegionDto> getRegions() {
        Map<Integer, List<RegionDto>> allRegion = regionDao.findAll().stream().map(RegionDto::new).collect(Collectors.groupingBy(RegionDto::getLevel));

        Map<String, RegionDto> cities = allRegion.get(2).stream().collect(Collectors.toMap(RegionDto::getValue, t -> t));
        Map<String, List<RegionDto>> work = allRegion.get(3).stream().collect(Collectors.groupingBy(RegionDto::getParent));
        work.forEach((p, rs) -> cities.get(p).setChildren(rs));

        Map<String, RegionDto> pronvinces = allRegion.get(1).stream().collect(Collectors.toMap(RegionDto::getValue, t -> t));
        work = allRegion.get(2).stream().collect(Collectors.groupingBy(RegionDto::getParent));
        work.forEach((p, rs) -> pronvinces.get(p).setChildren(rs));

        return allRegion.get(1);
    }

    private Map<String, Object> wraperFuncToMap(Function func) {
        Map<String, Object> f = new HashMap<>();
        f.put("id", func.getId().toString());
        f.put("name", func.getName());
        f.put("icon", func.getName());
        f.put("link", func.getName());
        if (Objects.nonNull(func.getParent())) {
            f.put("parent", func.getParent().getId());
        }
        f.put("visable", true);
        return f;
    }
}

File: test/src/main/kotlin/com/shipping/service/SystemService.kt
package com.shipping.service

import com.shipping.domain.dto.MenuDto
import com.shipping.domain.dto.RegionDto
import com.shipping.domain.enums.FunctionTypeE
import com.shipping.domain.sys.Function
import com.shipping.domain.sys.Region
import mu.KLogging
import org.spin.core.collection.FixedVector
import org.spin.core.session.SessionManager
import org.spin.core.util.MapUtils
import org.spin.data.extend.RepositoryContext
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.util.Objects
import kotlin.collections.ArrayList

/**
 *
 * Created by xuweinan on 2017/8/8.
 *
 * @author xuweinan
 */
@Service
class SystemService {
    companion object : KLogging()

    @Autowired
    private lateinit var repoCtx: RepositoryContext

    @Autowired
    private lateinit var userService: UserService

    /**
     * 组装菜单树
     */
    val menus: List<MenuDto>
        get() {
            val result = ArrayList<MenuDto>()

            val group = userService.getUserFunctions(SessionManager.getCurrentUser().id)[FunctionTypeE.MEMU]!!.map { MenuDto(it) }.groupBy { it.idPath.split(",").size }
//                .stream()
//                .map<MenuDto>(Function({ MenuDto.toDto(it) })).collect<Map<Int, List<MenuDto>>, Any>(Collectors.groupingBy { f -> f.idPath!!.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray().size })

            val maxDept = group.keys.stream().max { obj, anotherInteger -> obj.compareTo(anotherInteger) }.orElse(0)
            var currDept = 1
            val parent = FixedVector<MutableList<MenuDto>>(1)

            while (currDept <= maxDept) {
                group[currDept++]!!.forEach { d ->
                    parent.put(result)
                    d.idPath.split(",").map { it.toLong() }.forEach {
                        val p = parent.get().firstOrNull { m -> it == m.id }
                        if (null != p) {
                            parent.put(p.children)
                        } else {
                            parent.get().add(d)
                        }
                    }
                }
            }
            return result
        }

    val regions: List<RegionDto>
        get() {
            val allRegion = repoCtx.findAll(Region::class.java).map { ::RegionDto.call() }.groupBy { it.level!! }

            val cities = allRegion[2]!!.map { it.value!! to it }.toMap()
            var work = allRegion[3]!!.groupBy { it.parent }
            work.forEach { p, rs -> cities[p]!!.children = rs }

            val pronvinces = allRegion[1]!!.map { it.value!! to it }.toMap()
            work = allRegion[2]!!.groupBy { it.parent }
            work.forEach { p, rs -> pronvinces[p]?.children = rs }

            return allRegion[1]!!
        }

    private fun wraperFuncToMap(func: Function): Map<String, Any?> {
        val f: MutableMap<String, Any?> = MapUtils.ofMap("id", func.id?.toString(), "name", func.name, "icon", func.icon, "link", func.link, "visable", true)
        if (Objects.nonNull(func.parent)) {
            f["parent"] = func.parent!!.id
        }
        return f
    }
}
